kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,305 @@
+                    //TODO: specialize square case
+                    let exp = FoldingExp::Mul(exp.clone(), Box::new(exp2));
+                    Term { exp, sign, degree }
+                };
+                let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+                extract_terms(*e.clone(), terms, &f);
+            }
+        }
+        FoldingExp::Add(e1, e2) => {
+            extract_terms(*e1, terms, f);
+            extract_terms(*e2, terms, f);
+        }
+        FoldingExp::Sub(e1, e2) => {
+            extract_terms(*e1, terms, f);
+            let f = |t: Term<C>| f(-t);
+            let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+            extract_terms(*e2, terms, &f);
+        }
+        FoldingExp::Mul(e1, e2) => {
+            let mut temp = Vec::with_capacity(4);
+            extract_terms(*e1, &mut temp, f);
+            for term in temp {
+                let Term { exp, sign, degree } = term;
+                let exp = Box::new(exp);
+                let f = |t| {
+                    let Term {
+                        exp: exp2,
+                        sign: sign2,
+                        degree: degree2,
+                    } = t;
+                    let degree = &degree * &degree2;
+                    let sign = sign == sign2;
+                    let exp = FoldingExp::Mul(exp.clone(), Box::new(exp2));
+                    Term { exp, sign, degree }
+                };
+                let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+                extract_terms(*e2.clone(), terms, &f);
+            }
+        }
+        e @ FoldingExp::UnnormalizedLagrangeBasis(_) => {
+            let term = f(Term::new(e, Zero));
+            terms.push(term)
+        }
+        FoldingExp::Cache(id, e) => {
+            let f = |t| {
+                let Term { exp, sign, degree } = t;
+                let exp = FoldingExp::Cache(id, Box::new(exp));
+                Term { exp, sign, degree }
+            };
+            let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+            extract_terms(*e, terms, &f);
+        }
+    }
+}
+
+///simplyfies the expression, and separates it into terms of different degree y
+pub fn folding_expression<C: Copy, F: Fn(&FeatureFlag) -> bool>(
+    exp: &Expr<C>,
+    flag_resolver: &F,
+) -> IntegratedFoldingExpr<C> {
+    let simplied = simplify_expression(exp, flag_resolver);
+    let mut terms = vec![];
+    extract_terms(simplied, &mut terms, &|t| t);
+    let mut integrated = IntegratedFoldingExpr::default();
+    for term in terms.into_iter() {
+        let Term { exp, sign, degree } = term;
+        let t = (exp, sign);
+        match degree {
+            Degree::Zero => integrated.degree_0.push(t),
+            Degree::One => integrated.degree_1.push(t),
+            Degree::Two => integrated.degree_2.push(t),
+        }
+    }
+    integrated
+}
+fn simplify_expression<C: Copy, F: Fn(&FeatureFlag) -> bool>(
```

Comment:
Could you add tests please?

Reply:
What kind of test do you have in mind? it isn't very far from a `From` implementation, I suppose I can write the expected results for some inputs and assert they are equal to the actual output, something like that?

Reply:
Yeah, something similar at least to be able to see at a first glance how they should look like